rs/ledger_suite/icrc1/ledger/src/main.rs

```diff
@@ -225,11 +235,74 @@ fn post_upgrade(args: Option<LedgerArgument>) {
 
     PRE_UPGRADE_INSTRUCTIONS_CONSUMED.with(|n| *n.borrow_mut() = pre_upgrade_instructions_consumed);
 
+    set_ledger_state(LedgerState::Migrating(LedgerField::Allowances));
+    Access::with_ledger_mut(|ledger| {
+        ledger.clear_arrivals();
```

Comment:
Good catch! I fixed this and added a test that catches such scenario!

Reply:
Since we now have to clear partially migrated data in pre_upgrade in case of incomplete migration, we probably won't be able to allow migrating from V2 directly to V4 - we wouldn't know if we need to clear only balances or also allowances in case of incomplete migration. If we want to allow V2 to V4 migration, we will have to persist the migration progress.

Reply:
We can probably use ledger_version - in post upgrade we need store the version we were upgrading from.

Reply:
Good point! Using `ledger_version` sounds like a good approach. We could also check the number of entries in the heap data structures vs stable structures and make a decision based on that, as discussed in the ["Migration State"](https://docs.google.com/document/d/1v34PJ0sgYKIKKF2r4o_qhVaeSHqeUwhmR7XnuE5UB0o/edit?tab=t.0#heading=h.45iv9u2wilbu) section of the design doc. So in the `post_upgrade` of V3, we should only clear the entries from the stable structures if we have data there and in the heap data structures - if there is no data in the heap data structure, we consider the migration completed and do not clear the stable structures data structures.

Reply:
In https://github.com/dfinity/ic/pull/1513/commits/8867fa3b47b0232e8c0de05e9a2e98a2a2ffd65a I check the `ledger_version` and only start migration if it is smaller, otherwise mark the ledger as ready. I found an edge case that upgrade V3->V3 could clear stable allowances.